### ISSUE TITLE
Command line arguments for python scene

### ISSUE THREAD

**Issue by chrijopa** [sofa-framework/sofa#356]
I am currently running a convergence analysis using python scenes. Since the convergence analysis loads several settings from a .json file, it would be good to have a possibility to forward the filename of the file to the python scene via the command line. To my knowledge, this is not possible at the moment, please correct me if I am wrong and explain me your solution.

My first work around was to load from the same file (I call it outputJson, see the function definition below). However, this becomes cumbersome, when the code is parallelized, since several threads would like to load from that same file.

My approach for the parallelization is to write the filename (which is now adapted to every scene) first into the .scn using PHP and then to load it from the python scene. 
That means, in the .scn file we have:
`<?php echo '<PythonScriptController classname="convergenceBeam_Template" filename="convergenceBeam_TemplatePython.py" listening="1" name="convergenceBeam_Template" variables="'.$argv[1].'" />'; ?>`
which called from the convergence python script and deleted later
```
os.system("php convergenceBeam_TemplatePython.scn "+sofaOutput+".pyscndata > "+curTime+".scn")
... call the scene
os.system("rm "+curTime+".scn")
```
In the .py file we load the data using
`self.defineDataFromJson(self.findData("variables").value[0][0])`
where
```
def defineDataFromJson(self,jsonFilename="outputJson") :
        f = open(jsonFilename,'r')
        self.data =  json.load(f)
        f.close()
def getValueOfData(self,keyOfData) :
        if keyOfData in self.data.keys():
            return self.data[keyOfData]
        printStack()
        sys.exit("\nERROR\n the key "+keyOfData+" is not defined!\nERROR\n")
```

Both of the solutions I consider as hacks, the first as it can not run in parallel and the second as the scene does not work without running php on the side, thus it does not work with one command line.

Potential (final?) solutions would be:
- a command line argument or command line argument**s** for python scenes
- a command line argument that can be used in all scenes, such as -d for data, which could be followed by strings
- php scenes (file ending .pscn) that work and could prevent calling php on the side (i.e. my second solution works without a script on the side)

I am grateful for any ideas/other propositions of solutions!

Label propositions: project: SofaPython, discussion

**Comment by maxime-tournier** [comment]
I most definitely agree with you, this is very annoying. In fact, I have a custom build of the main SOFA executable just for this.

As of now, there is no mechanism I know of that will forward `sys.argv` correctly. The simplest and by far the most natural solution would be to forward anything happening after `--` to `sys.argv`. 

Unless there is a compelling reason not to do so (@damienmarchal ?), I'll give it a quick shot soon.

**Comment by maxime-tournier** [comment]
Ok so after a quick look at the [code](https://github.com/sofa-framework/sofa/blob/master/SofaKernel/framework/sofa/helper/ArgumentParser.cpp#L128), there is already a path that handles `--` in a non-standard way.

Any argument following `--` is added to the files list, and all but the first files are ignored in the main `runSofa` executable. I don't really know if there are other binaries relying on this behavior, so what we could do is:
- either handle `--` properly, i.e. all arguments after `--` are ignored by SOFA and made available for further use, in this case by forwarding them to python.
- or add an `--extra` or `--argv` special argument that will do the same without interfering with current behaviour.

**Comment by damienmarchal** [comment]
Hi all,

To me there is several questions here... 
1- one is how to pass command line argument to python.
2- one is how to get the file name associated with a loaded scene.
3- one is how to execute multiple sofa in // with small difference in the scene.
 
1-
To pass command line argument to python and answering Maxime suggestion I prefer something like --extra or --scene-argv so that we don't have to take care of not breaking someone's code. 

2-
Getting the filename using a binded function in the Sofa python module.
Filename only make sense in specific case, i.e. when we have a single scene loaded from a file (as in runSofa) of course this means nothing when you have no scene files (dynamically generating the scene) or you have loaded multiple scene. So I think this shouldn't be part of the Sofa python module but could find its place in a separated RunSofa python module that would expose things specific to the runSofaX application. Actually I would like to have this kind of module for other reasons. 

3-
About the general workflow of launching multiple scene. We have in Defrost few python scripts called sofa-launcher that we use to execute in parallel multiple sofa instances of a templated scene. The parallel execution can be done on a single machine or a a cluster of linux with ssh access, scene specific data are saved in temporary directories or gathered by the calling script for further processing (eg: plotting the result). think this could prefectly match what you want to do so I will make a PR for that.

**Comment by maxime-tournier** [comment]
Hi, thanks four your feedback :-)

I'm tackling 1. and it's almost done, however I lack a clean way of exposing the extra args to the sofa python plugin. any idea ?
2. can't you just use `__file__` for that anyway ?

**Comment by damienmarchal** [comment]
Hi Maxime,

For 1. 
What about just setting the arguments into sys.argv using PySys_SetArgv(argc, argv) ? 

For 2. 
Isn't __ file __ returning the current file... So in a python script controller __ file __ it would not contain the scene filename but the controller's one. But you are totally right that this could solve the initial @chrijopa question in a very simple way.

**Comment by maxime-tournier** [comment]
Mostly done, see #368

**Comment by chrijopa** [comment]
Thank you a lot for the nice proposals preventing potential hacks like the ones I presented above. I just checked your pull request and I really like its simple usage - I hope it will be merged soon! 

@damienmarchal : I am excited to see your solution to 3., if you would like to have further information about my approach, please let me know.

To create python scenes, I am using the [scene to python transform script](https://github.com/sofa-framework/sofa/blob/master/applications/plugins/SofaPython/scn2python.py) and with the direct usage your approach works well, i.e.
`applications/plugins/SofaPython/scn2python.py examples/Demos/caduceus.scn -o caduceus2 -p`
generates the python scene `caduceus2.py` and when launching it I can retrieve all the arguments of the command line, i.e. the output of 
`runSofa caduceusPython.scn --argv test test2`
is `['caduceusPython', 'test', 'test2']`.

However, by default (i.e. without -p) the scene to python transform script generates a .scn file calling a .py file - a choice made to prevent having to load the python plugin before launching python scenes. For example
`applications/plugins/SofaPython/scn2python.py examples/Demos/caduceus.scn`
outputs the caduceusPython.scn and caduceusPython.py. Now when using 
`runSofa caduceusPython.scn --argv test test2`
and running 
`print sys.argv`
only the filename can be retrieved, i.e. the output is `['caduceusPython']`.

1. Is the usage presented as second approach (i.e. using .scn and .py) still up to date/a recommended approach?
2. If yes, is it possible to access the argv in the .py file when it is used from the .scn file? Or is it possible to improve the implementation such that it can handle it?

**Comment by maxime-tournier** [comment]
Hi there,

Unless you really have to, I would suggest you drop `.scn` files altogether and work with python files directly: it's easier to work with and waaay more powerful. If it were just me, I'd have dropped `.scn` support a long time ago or rewritten it with 100 lines of python :-) Some people may disagree though.

I'm not sure there is a nice way of forwarding command line args from the `.scn` scene to the python script. It seems possible though, but I don't want to mess with existing code passing arguments to python scripts as it may break older scenes.

**Comment by chrijopa** [comment]
As a reaction to your proposition I adapted my scene and dropped the .scn file and now it works, thank you for your work! 

PS (another issue?): Since createScene does not have the argument self - that I am using to communicate values between the different functions - I inserted a python script controller, i.e. 
```python
def createScene(rootNode) :
    rootNode.createObject('PythonScriptController',classname="convergenceBeam_Template", 
                          filename="convergenceBeam_TemplatePython.py", ... , 
                          name="convergenceBeam_Template", variables=sys.argv[1]+" "+sys.argv[2])
```
and 
```python
class convergenceBeam_Template (Sofa.PythonScriptController):
    def createGraph(self,rootNode):
        self.defineDataFromJson(self.findData("variables").value[0][0])
        self.outputJsonAtTimeStep = ast.literal_eval(self.findData("variables").value[1][0])
```
This does not seem to be very elegant.
1. Could self be used as well as an argument for createScene and createSceneAndController?
2. Is there another way to communicate values between the functions (e.g. createScene and onBeginAnimationStep)?

If there is none, I am open to write another issue (I discussed with a colleague of mine and he had the same problem before)!

[@maxime-tournier edit: fixed formatting]

**Comment by maxime-tournier** [comment]
> 1. Could self be used as well as an argument for createScene and createSceneAndController?

It's unclear what you're trying to do: `createScene` is a function and has no `self` binding.

> 2. Is there another way to communicate values between the functions (e.g. createScene and onBeginAnimationStep)?

I think #283 was merged, so what you can do now is the following:

```python
import Sofa

class MyScript(Sofa.PythonScriptController):

    def __init__(self, node, foo):
        # note: the node argument is mandatory
        self.foo = foo
        
    def onBeginAnimationStep(self, dt):
        print(self.foo)
        
def createScene(node):
    foo = 10
    my_script = MyScript(node, foo)
```

**Comment by chrijopa** [comment]
Thank you for your help! I changed the scn2python script:
- The standard output is now be .py (before it was .scn and .py)
- It handles the command line arguments now in the way you presented it above
- It uses the output into .json, as presented in #360.
I would like to push these changes to your pull requestion #368, do you mind giving me access rights (if not I will wait and do a separate pull request or I push it to #360 (which is not as close as your pull request) )?

**Comment by maxime-tournier** [comment]
Please make a separate pull request referencing #368. 

I need to fix failing tests before #368 is merged though.

**Comment by ghost** [comment]
Sounds good, I will check regularly to see whether your pull request is finished!

**Comment by maxime-tournier** [comment]
#368 is merged, closing issue.

**Comment by maxime-tournier** [comment]
#368 is merged, closing issue.
